src/workerd/io/worker-fs.h

```diff
@@ -0,0 +1,638 @@
+  // the recursive option is not set, an exception will be thrown if the directory
+  // is not empty. If the directory is read only, an exception will be thrown.
+  // If the node is a file, it will be removed regardless of the recursive option
+  // if the directory is not read only.
+  // If the node is a symlink, the symlink will be removed but the target will
+  // be left intact.
+  // The path must be relative to the current directory.
+  // Note that this method will only remove the node from this directory. If the
+  // node as references in other directories, those will not be removed.
+  // If the target is a symbolic link, the symbolic link will be removed but
+  // the target will not be removed.
+  virtual bool remove(
+      jsg::Lock& js, kj::PathPtr path, kj::Maybe<RemoveOptions> options = kj::none) = 0;
+
+  virtual kj::StringPtr jsgGetMemoryName() const = 0;
+  virtual size_t jsgGetMemorySelfSize() const = 0;
+  virtual void jsgGetMemoryInfo(jsg::MemoryTracker& tracker) const = 0;
+
+  // Creates a new readable/writable in-memory directory. This directory will
+  // not be initially included in a directory. To add it to a directory, use the
+  // add method. If the directory is not added, it will be deleted
+  // when the handle is dropped.
+  static kj::Rc<Directory> newWritable();
+
+  // Used to build a new read-only directory. All files and directories added
+  // may or may not be writable. The directory will not be initially included
+  // in a directory. To add it to a directory, use the add method.
+  class Builder {
+   public:
+    Builder() = default;
+    KJ_DISALLOW_COPY_AND_MOVE(Builder);
+
+    // Adds a file or directory to the directory. The name must not contain any
+    // path separators. If the name already exists an exception is thrown.
+    void add(kj::StringPtr name, kj::OneOf<kj::Rc<File>, kj::Rc<Directory>> fileOrDirectory);
+
+    // Adds a directory builder to the directory. This is used to incrementally build
+    // up read-only directories. When finish is called, the directory will be
+    // finalized. The name must not contain any path separators.
+    void add(kj::StringPtr name, kj::Own<Directory::Builder> dir);
+
+    // Adds a file or directory at the given path. The path may contain path separators.
+    // Subdirectories will be created as needed (as read-only directories).
+    void addPath(kj::PathPtr path, kj::OneOf<kj::Rc<File>, kj::Rc<Directory>> fileOrDirectory);
+
+    // Finalizes and returns the directory. The directory will not be writable.
+    // The directory will not be initially included in a directory. To add it
+    // to a directory, use the add method.
+    kj::Rc<Directory> finish();
+
+    using Map = kj::HashMap<kj::String,
+        kj::OneOf<kj::Rc<File>, kj::Rc<Directory>, kj::Own<Directory::Builder>>>;
+    using Entry = Map::Entry;
+
+   private:
+    Map entries;
+  };
+};
+
+// The equivalent to a symbolic link. A symlink holds a reference to both the
```

Comment:
Nit:
The comment here could be clarified as the symlink does *not* hold a reference to the target node, it just holds its path.
This mimics linux' behaviour under which the target node can be deleted and recreated and the symlink will still be valid.

Reply:
The comment here is that the symlink holds a reference to the VirtualFileSystem and a target path, not that it holds a reference to the target node.